[5주차] 이경준/[feat] 로그인&회원가입 api 연동

KyeongJooni opened the pull request:
## ✨ 주요 변경사항
- 토큰 관리 로직 중앙화
- axios 인터셉터 활성화
- 카카오 OAuth 로그인 구현
- 회원가입 플로우 개선

---

## 📝 작업 상세 내용
1. API 인터셉터 구현
- `axios-auth-refresh` 라이브러리 도입으로 토큰 재발급 자동화
- 요청 인터셉터: Authorization 헤더 자동 추가
- 응답 인터셉터: body의 `code` 값 기반 에러 처리
- 카카오 콜백 API는 인터셉터 예외 처리 (401이 회원가입 필요를 의미)

2. 카카오 로그인
- `useMutation` 대신 Promise로 직접 처리하여 코드 간소화
- response code 기반 분기:
  - `200`: 기존 회원 → 자동 로그인 후 홈 이동
  - `401`: 신규 회원 → 회원가입 페이지 이동
- sessionStorage를 통한 카카오 회원가입 정보 관리

3. 타입 정의 개선
- `ApiResponse<T>` 공통 타입 추가로 API 응답 타입 일관성 확보
- 인증 관련 타입 세분화 및 정리

---

## ✅ 체크리스트
- [x] GitHub 이슈에 `resolves: #번호`를 추가했는가?
- [x] 불필요한 주석, 디버깅 코드 제거했는가?
- [x] 기능 테스트 및 정상 동작 확인했는가?

---

## 📸 스크린샷 (선택)
<img width="1919" height="1018" alt="image" src="https://github.com/user-attachments/assets/0cb4257d-46e1-496c-8585-0eb2b07a993c" />

일반 회원가입/로그인

<img width="1919" height="1016" alt="image" src="https://github.com/user-attachments/assets/ba2d86f9-b402-4237-a087-96fa448e91b5" />

카카오 회원가입/로그인

---

## 🔍 기타 참고사항
- 부랴부랴 만드느라고 코드가 엉망진창입니다. 리뷰 달게 받겠습니다...
- 아직 구현할 부분이 산더미입니다 ㅠㅠ

#### 추후 개선 필요 사항
- 에러 바운더리 추가
- 로딩/에러 상태에 대한 전역 처리

---

## 🔗 관련 이슈
- Close #43

KyeongJooni committed:
feat: 인증 API 타입 시스템 및 React Query 통합

- 실제 API 스펙에 맞춰 인증 관련 타입 정의
  - 로그인/회원가입/OAuth/토큰재발급 타입 구현
  - BaseUserInfo 공통 타입으로 중복 제거
  - ApiResponse<T> 제네릭 패턴으로 Response 타입 제거
  - 네임스페이스 패턴(AuthTypes)으로 import 간소화
- React Query 설치 및 mutation 훅 구현
  - useRegisterMutation, useLoginMutation 등
- useSignup 훅에서 회원가입 API 연동 완료

KyeongJooni committed:
feat: 로그인/회원가입 API 연동 및 사용자 정보 조회 구현

KyeongJooni committed:
refactor: 토큰 관리 함수를 apiInstance로 중앙화

- getAccessToken, setAccessToken 함수 추가
- getRefreshToken, setRefreshToken 함수 추가
- authApi와 useAuthStore에서 중복 로직 제거
- 토큰 관리를 apiInstance에서 일괄 처리

KyeongJooni committed:
feat: 공통 API 타입 정의 추가

- ApiResponse 공통 타입 추가
- 타입 export 경로 정리

KyeongJooni committed:
refactor: 카카오 로그인을 Promise 방식으로 리팩토링

- useKakaoCallback을 Promise 기반으로 단순화
- 카카오 로그인 백엔드 엔드포인트로 직접 리다이렉트
- 카카오 콜백 페이지 UI 개선
- 인증 타입 정의 개선

KyeongJooni committed:
refactor: 회원가입 로직 및 폼 개선

- 카카오 회원가입 플로우 개선
- sessionStorage 기반 카카오 정보 관리
- 회원가입 폼 필드 정리
- validation 스키마 개선

KyeongJooni committed:
refactor: 유저 API 타입 정의 개선

- ApiResponse 타입 적용
- 타입 안정성 향상

KyeongJooni committed:
refactor: 레이아웃 및 라우팅 개선

- Layout 컴포넌트 인증 체크 로직 개선
- App 라우팅 정리

KyeongJooni committed:
chore: 의존성 업데이트 및 설정 개선

- axios-auth-refresh 추가
- gitignore 업데이트

KyeongJooni committed:
fix: 로그인 폼 브라우저 경고 제거

- form 태그로 감싸서 password input 경고 해결
- autocomplete='off' 속성 추가하여 자동완성 경고 해결
- Enter 키로 로그인 가능하도록 onSubmit 핸들러 추가

KyeongJooni committed:
feat: 카카오 OAuth 응답 코드 및 리다이렉트 경로 상수 추가

- KAKAO_RESPONSE_CODE: 응답 코드 상수화
- KAKAO_REDIRECT_PATH: 리다이렉트 경로 상수화

KyeongJooni committed:
refactor: useKakaoCallbackMutation에 옵션 파라미터 지원 추가

- onSuccess, onError 콜백을 훅 레벨에서 전달 가능하도록 개선
- React Query v5에서 비동기 콜백 처리 안정성 향상

KyeongJooni committed:
refactor: useKakaoCallback 훅 리팩토링

- Promise 방식에서 useMutation으로 변경
- 응답 처리 로직을 함수로 분리 (handleExistingUser, handleNewUser)
- 상수를 oauth.constants.ts로 이동
- window.location.href를 navigate로 변경하여 SPA 성능 개선
- URL 파라미터 추출 및 sessionStorage 저장 로직 함수로 분리

KyeongJooni committed:
refactor: 컴포넌트 파일명 및 타입 일관성 개선

- Variants 파일명을 PascalCase로 통일
  - textFieldVariants.ts -> TextFieldVariants.ts
  - textBoxVariants.ts -> TextBoxVariants.ts
  - buttonVariants.ts -> ButtonVariants.ts
  - modalVariants.ts -> ModalVariants.ts
  - paginationButtonVariants.ts -> PaginationButtonVariants.ts

- textbox 디렉토리를 TextBox로 변경하여 컴포넌트명과 일치

- TextField 컴포넌트 개선
  - BaseInputProps에 autoComplete 타입 추가
  - TextField에 autoComplete 속성 추가 (기본값: 'off')
  - 브라우저 autocomplete 경고 제거

KyeongJooni committed:
fix: 타입 오류 및 코드 스타일 수정

KyeongJooni committed:
refactor: 불필요한 주석 제거

rename: 폴더명 대문자 변경

KyeongJooni committed:
rename: 폴더명 파스칼케이스 변경

KyeongJooni committed:
refactor: authQuery mutation을 React Query 표준 패턴으로 개선

- useRegisterMutation, useLoginMutation 등에 queryClient 추가
- mutation 성공 시 userInfo 쿼리 자동 무효화
- useKakaoCallbackMutation만 옵션 파라미터 유지 (복잡한 비즈니스 로직)
- React Query의 onSuccess 콜백으로 캐시 관리 일관성 확보

KyeongJooni committed:
feat: useAuth 헬퍼 훅 추가

- useUserInfo 기반으로 인증 상태를 제공하는 useAuth 헬퍼 추가
- user, isLoggedIn, isLoading, error를 통합 제공
- React Query 데이터를 기반으로 계산된 값 반환

KyeongJooni committed:
refactor: authStore를 UI 상태 관리로 간소화

KyeongJooni committed:
refactor: 인증 관련 훅을 React Query 기반으로 리팩토링

KyeongJooni committed:
refactor: Route 컴포넌트를 useAuth 훅 기반으로 변경

KyeongJooni committed:
refactor: 레이아웃 및 컴포넌트를 useAuth 기반으로 마이그레이션

KyeongJooni committed:
mod: isLoggedIn만 필요한 컴포넌트 Zustand로 최적화

KyeongJooni committed:
refactor: 불필요한 주석 제거

eileen4505 commented on src/utils/validation.ts:
```diff
@@ -28,6 +30,7 @@ export const VALIDATION_MESSAGES = {
+    format: '닉네임은 영문, 숫자만 가능합니다.',
   },
   bio: {
     maxLength: '한 줄 소개는 최대 30글자입니다.',
```
입력 검증 부분을 어떻게 해야하나 고민이었는데, 배워갑니다!

eileen4505 commented on src/api/user/userQuery.ts:
```diff
@@ -0,0 +1,42 @@
+    isLoading,
+    error,
+  };
+};
```
userInfo를 react-query 단일 소스로 두고, Zustand에는 로그인 여부만 둔게 불필요한 중복 없이 정말 좋은 것 같아요!

eileen4505 reviewed (approved):
항상 구현하신 코드 보면서 많이 배워가는 것 같아요! 고생 많으셨습니다👍🏻

gachaemin commented on src/api/auth/authApi.ts:
```diff
@@ -1,82 +1,66 @@
 // 로그인 API
-export const login = async (loginData: LoginRequest): Promise<LoginResponse> => {
-  const response = await axiosInstance.post<LoginResponse>('/auth/login', loginData);
+export const login = async (loginData: AuthTypes.LoginRequest): Promise<ApiResponse<AuthTypes.LoginData>> => {
```
`apiResponse<T>` 제네릭 타입을 사용하면 응답 구조를 통일할 수 있군요! 이를 사용하면 구조를 매번 확인하지 않아도 될거 같아요!

gachaemin reviewed (commented):
항상 너무 잘하셔서 남길 코멘트가 없습니다... 수고하셨어요!

ld5n1z commented:
경준님 5주차 과제 진행하시느라 수고 많으셨습니다!!! 👏

HandoA01 commented:
5주차 개발 수고 많으셨습니다! 구현하신 코드에서 배울점이 많은 것 같아요 ㅎㅎ 많이 배워갑니다 :)

ld5n1z commented on src/api/auth/authQuery.ts:
보통 Query파일을 분리하는 편이 좋을까요? 
분리한다면 어느 기준에서 분리하는지 궁금합니다!!!

KyeongJooni replied:
프로젝트의 단위에 따라  보통 분리하는거에 시간을 많이 투자하는 것 같아요 다양한 설계 방식이 있는거로 알고 있긴 한데 대부분 여태까지 했던 프로젝트가 완전히 대형 서비스는 아니여서... API가 산더미처럼 많아지면 Mutation을 Hooks에 넣어서 관리하게 되더라도 지저분해진다고 생각해요 그래서 보통 보편적인 동아리 프로젝트 규모정도에서는 전역에 query폴더를 만들어서 React Query만 관리하거나 역할 별로 묶어서 관리하려고 하는 편이에요!

ld5n1z replied:
답변 감사합니다!!! 👍👍

ld5n1z commented on src/api/user/userQuery.ts:
```diff
@@ -0,0 +1,42 @@
+    refetchOnWindowFocus: false,
+  });
+
+  useEffect(() => {
```
React Query의 데이터/에러 상태를 감지해서, Zustand의 isLoggedIn을 즉시 반영한다는 점 배워갑니다 🙌

HandoA01 commented on src/api/user/userQuery.ts:
```diff
@@ -0,0 +1,42 @@
+export const useUserInfo = () => {
+  const setIsLoggedIn = useAuthStore(state => state.setIsLoggedIn);
+
+  const query = useQuery<ApiResponse<UserTypes.UserData>>({
+    queryKey: ['userInfo'],
+    queryFn: getUserInfo,
+    staleTime: 5 * 60 * 1000, // 5분
+    gcTime: 10 * 60 * 1000, // 10분
+    retry: false, // 401 에러 시 재시도하지 않음
+    refetchOnWindowFocus: false,
+  });
```
Query와 Api, Type 등 코드가 분리 되어 있어 가독성이 좋은것 같아요!!

theSnackOverflow commented:
경준님은 매주 과제를 완벽하게 수행하셔서 과제 리뷰보다는 코드를 보며 오히려 많이 배우는 것 같습니다..!
이번주 과제도 수행하시느라 고생 많으셨습니다!

KyeongJooni committed:
remove: 불필요한 파일 삭제

dalzzy commented on src/api/auth/authQuery.ts:
```diff
@@ -1 +1,53 @@
+import type { ApiResponse } from '../apiTypes';
+import type * as AuthTypes from './authTypes';
+
+export const useRegisterMutation = () => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: register,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['userInfo'] });
+    },
+  });
+};
+
+export const useRegisterOAuthMutation = () => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: registerOAuth,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['userInfo'] });
+    },
+  });
+};
+
+export const useLoginMutation = () => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: login,
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ['userInfo'] });
+    },
+  });
+};
```
useQueryClient를 매번 호출해도 기능상 문제는 없지만, 코드 중복을 방지하기 위해 별도로 분리해서 사용하는 게 어떨까요?!

저는 보통 lib 폴더에 이런 식으로 정의해두는 편,,, 입니당,, 참고만 해주시면 좋을 것 같아욤
```
// lib/queryClient.ts
import { QueryClient } from '@tanstack/react-query';

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      refetchOnMount: false,
    },
  },
});

```

KyeongJooni replied:
별도로 분리해서 사용하는 방법도 있군요!  저는 각 hook에서 useQueryClient()를 사용하는 건 중복이 아니라 의존성을 명시적으로 선언하는 React 표준 패턴인거로 알고 있어서 저렇게 사용했습니다! 근데 만약 분리를 하게 된다면 분리한 lib에서 export한 queryClient와 App.tsx Provider의 queryClient가 달라 캐시 동기화가 안 되는 상황이 있지는 않은지 조심스레 여쭤봅니다...

dalzzy replied:
엇 `App.tsx` 에서도 새로운 QueryClient 인스턴스를 생성해서 사용하는게 아니라 lib폴더에 분리한 queryClient를 import해서 사용하도록 수정하면, 전역에서 동일한 인스턴스를 공유하게 돼서 괜찮지 않을까욤...?...

KyeongJooni replied:
저도 useQueryClient 훅을 명시해서 사용하는뿐이 안해봐서 잘 몰라서 찾아봤는데, 공식 문서에서도 기본 패턴으로 useQueryClient() 훅 사용을 제시했더라구요! 추가로 AI 끄적이면서 더 찾아보니까 Next로 SSR 사용시에는 데이터 누수 가능성도 있다고 뭐 엄청 설명하더라구요... 참고만 해보시면 될 것 같습니다!

dalzzy replied:
아하 그러쿤요 😲 그럼 지금 방식으로 적용해도 괜찮을 것 같네요!! 몰랐던 부분인데 답변 감사합니당,,

dalzzy commented on src/api/apiInstance.ts:
```diff
@@ -1,48 +1,48 @@
     return null;
   }
-  return localStorage.getItem('accessToken');
+  return sessionStorage.getItem('accessToken');
```
로컬스토리지에서 세션스토리지로 바꾸신 이유가 궁금합니당!-!

KyeongJooni replied:
토큰이 브라우저를 닫으면 자동으로 만료되도록 하고 싶어서 sessionStorage로 변경했습니다!
로그인 상태를 영구적으로 유지하기보다는, 하루 단위 세션 개념에 맞게 관리하려는 의도입니당 🙂

dalzzy commented on src/api/auth/authTypes.ts:
BaseUserInfo 의 일부가 여러 타입에 반복되고 있는 것 같은데 `pick`이나 `omit`을 활용해서 필요한 속성만 재활용하는 건 어떨지 제안드려 봅니다,,!! 

예를 들면, TokenPair, BaseUserInfo 타입을 활용해서 
RegisterData의 경우,
```
export type RegisterData = Pick<BaseUserInfo, 'email' | 'nickname' | 'profilePicture' | 'introduction'>;
```
이런 식으로 쓸 수 있을 것 같아요! 

추가로 KakaoCallbackData 는 
```
export type KakaoCallbackData = Partial<TokenPair> &
  Partial<BaseUserInfo> & {
    httpStatus: string;
    responseMessage: string;
    kakaoId?: number;
  };
```
이런 식으로 단순화할 수 있을 것 같습니다!!

그리고 LoginData의 경우,TokenPair + nickname, profilePicture, introduction + httpStatus, responseMessage 형태니까
아래처럼 BasicProfile 타입을 한 번 만들어두면 이렇게 정리할 수 있을 것 같아욤

```
type BasicProfile = Pick<BaseUserInfo, 'nickname' | 'profilePicture' | 'introduction'>;

export interface LoginData extends TokenPair, BasicProfile {
  httpStatus: string;
  responseMessage: string;
}
```

그냥 제 개인적인 의견입니당......

KyeongJooni replied:
좋은 방향 같아요! 말씀하신 대로 Pick, Partial을 활용해서 타입 중복을 제거했습니다!!
코드가 훨씬 깔끔해지고 유지보수하기 더 좋아진 것 같아요!! 👍

dalzzy reviewed (approved):
5주차도 수고하셨습니다~!! 😁 제가 감히 리뷰할 것이 없 습니 다 ,, 👍🏻 코멘트 간단하게만 남겼으니 확인 하시고 답변 남겨주세용!! 
파이팅 ~~~

KyeongJooni committed:
refactor: authTypes 타입 중복 제거

- BasicProfile 유틸리티 타입 추가
- LoginData, RegisterData, KakaoCallbackData에 Pick/Partial 적용
- BaseUserInfo 재사용
